tool/tsh/common/db_exec.go

```diff
@@ -0,0 +1,647 @@
+	case byNames && bySearch:
+		return trace.BadParameter("--labels/--search flags cannot be used with --dbs flag")
+	}
+
+	// Logging.
+	if cf.ParallelJobs > 1 && cf.OutputDir == "" {
+		return trace.BadParameter("--output-dir must be set when executing concurrent connections")
+	}
+	if cf.OutputDir != "" && utils.FileExists(cf.OutputDir) {
+		return trace.BadParameter("directory %q already exists", cf.OutputDir)
+	}
+	return nil
+}
+
+func (c *databaseExecCommand) getDatabases() ([]types.Database, error) {
+	if c.cf.DatabaseServices != "" {
+		return c.getDatabasesByNames()
+	}
+	return c.searchDatabases()
+}
+
+func (c *databaseExecCommand) getDatabasesByNames() ([]types.Database, error) {
+	// Use a single predicate to search multiple names in one shot but batch 100
+	// names at a time. Extra validation will be performed afterward to ensure
+	// we fetched what we need.
+	fmt.Fprintln(c.cf.Stdout(), "Fetching databases by name ...")
+
+	// Trim spaces.
+	names := strings.Split(c.cf.DatabaseServices, ",")
+	for i := range names {
+		names[i] = strings.TrimSpace(names[i])
+	}
+	names = apiutils.Deduplicate(names)
+
+	var dbs []types.Database
+	for page := range slices.Chunk(names, 100) {
+		var predicate string
+		for _, name := range page {
+			predicate = makePredicateDisjunction(predicate, makeDiscoveredNameOrNamePredicate(name))
+		}
+
+		logger.DebugContext(c.cf.Context, "Getting database by name", "databases", page)
+		pageDBs, err := c.client.listDatabasesWithFilter(c.cf.Context, &proto.ListResourcesRequest{
+			Namespace:           apidefaults.Namespace,
+			ResourceType:        types.KindDatabaseServer,
+			PredicateExpression: predicate,
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		dbs = append(dbs, pageDBs...)
+	}
+
+	logger.DebugContext(c.cf.Context, "Fetched databases by name",
+		"databases", logutils.IterAttr(types.ResourceNames(dbs)))
+	return dbs, trace.Wrap(ensureEachDatabase(names, dbs))
+}
+
+func (c *databaseExecCommand) searchDatabases() (databases []types.Database, err error) {
+	fmt.Fprintln(c.cf.Stdout(), "Searching databases ...")
+	filter := c.tc.ResourceFilter(types.KindDatabaseServer)
+
+	logger.DebugContext(c.cf.Context, "Searching for databases", "filter", filter)
+	dbs, err := c.client.listDatabasesWithFilter(c.cf.Context, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger.DebugContext(c.cf.Context, "Fetched databases with search filter",
+		"databases", logutils.IterAttr(types.ResourceNames(dbs)),
+	)
+	return dbs, trace.Wrap(c.printResultAndConfirm(dbs))
+}
+
+func (c *databaseExecCommand) printResultAndConfirm(dbs []types.Database) error {
+	if len(dbs) == 0 {
+		return trace.NotFound("no databases found")
+	}
+
+	fmt.Fprintf(c.cf.Stdout(), "Found %d database(s):\n\n", len(dbs))
+	printTableForDatabaseExec(c.cf.Stdout(), dbs)
+	question := fmt.Sprintf("Do you want to proceed with %d databases?", len(dbs))
```

Comment:
Nit: To keep it simple and make the sentence work when there's only 1.
```suggestion
	question := fmt.Sprintf("Do you want to proceed with %d database(-s)?", len(dbs))
```